### PR TITLE
Move heartbeat user-reporting update to async staging pipeline

### PR DESCRIPTION
**Pull request by sravfeyn**
Implements https://dimagi-dev.atlassian.net/browse/IIO-642  https://github.com/dimagi/commcare-hq/pull/26001#issuecomment-557137784

### CODE REVIEW

File: corehq/pillows/synclog.py [hunk 1]
```diff
@@ -100,23 +100,24 @@ def process_change(self, change):
...
-            mark_last_synclog(domain, user, app_id, build_id, sync_date, device_id)
 
+            device_app_meta = None
+            if device_id and app_id:
+                device_app_meta = DeviceAppMeta(app_id=app_id, build_id=build_id, last_sync=sync_date)
+            mark_last_synclog(domain, user, app_id, build_id, sync_date, device_id, device_app_meta)
 
-def mark_last_synclog(domain, user, app_id, build_id, sync_date, device_id, save_user=True):
+
+def mark_last_synclog(domain, user, app_id, build_id, sync_date, device_id, device_app_meta, build_profile_id=None, save_user=True):
```
Comment (mkangia) [inline corehq/pillows/synclog.py]: not clear about the changes to synclog, why is this being done here?

Comment (sravfeyn) [reply corehq/pillows/synclog.py]: What's happening in the heartbeat view [above](https://github.com/dimagi/commcare-hq/pull/26305/files#diff-291dfd3802bf0daa51e598788d5ca13bL352) is pretty much same as what's happening inside `mark_last_synclog`, so I am modifying it to make this more DRY

Comment (mkangia) [reply corehq/pillows/synclog.py]: i am not sure about other fields but `build_profile_id` is only present in heartbeat requests and not in sync log, so will this clear out the value set for it via heartbeat?

Comment (sravfeyn) [reply corehq/pillows/synclog.py]: Each action just updates the relevant fields doesn't override other fields. See `add_sync` method [here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/users/models.py#L2850) which doesn't override that field.

Comment (mkangia) [reply corehq/pillows/synclog.py]: I am not sure if that link took me to where you expected me to see,
it was on the `ON CONFLICT (domain, user_id, app_id)` line in the sql,
is that what you were pointing to?

the github links with master as reference would keep changing, i find it useful to have them based on a commit which would not change.

Comment (sravfeyn) [reply corehq/pillows/synclog.py]: Yeah.

**Comment by sravfeyn** [comment]
Also @snopoke

File: corehq/apps/users/models.py [hunk 2]
```diff
@@ -2856,5 +2864,51 @@ def add_sync(cls, domain, user_id, app_id, build_id, sync_date, device_id):
...
+                    build_id = EXCLUDED.build_id,
+                    sync_date = EXCLUDED.sync_date,
+                    device_id = EXCLUDED.device_id,
+                    app_version = EXCLUDED.app_version,
+                    num_unsent_forms = EXCLUDED.num_unsent_forms,
+                    num_quarantined_forms = EXCLUDED.num_quarantined_forms,
+                    commcare_version = EXCLUDED.commcare_version,
+                    build_profile_id = EXCLUDED.build_profile_id,
+                    last_heartbeat = CLOCK_TIMESTAMP()
+                WHERE staging.sync_date IS NULL OR EXCLUDED.sync_date > staging.sync_date
```
Comment (snopoke) [inline corehq/apps/users/models.py]: shouldn't this be `WHERE staging.last_heartbeat is NULL or EXCLUDED.last_heartbeat > staging.last_heartbeat`

File: testapps/test_pillowtop/tests/test_xform_pillow.py [hunk 3]
```diff
@@ -161,6 +162,46 @@ def test_app_metadata_tracker_synclog_processed(self):
...
+        self.assertEqual(len(ccuser.reporting_metadata.last_submissions), num_submissions)
+        self.assertEqual(len(ccuser.reporting_metadata.last_syncs), 1)
+        self.assertEqual(ccuser.reporting_metadata.last_syncs[0].sync_date, sync_date)
+        self.assertEqual(ccuser.reporting_metadata.last_sync_for_user.sync_date, sync_date)
+        self.assertEqual(ccuser.last_device.device_id, 'heartbeat_device_id')
+        app_meta = ccuser.last_device.get_last_used_app_meta()
+        self.assertEqual(app_meta.num_unsent_forms, 2)
+        self.assertEqual(app_meta.num_quarantined_forms, 10)
+
+    @run_with_all_backends
```
Comment (snopoke) [inline testapps/test_pillowtop/tests/test_xform_pillow.py]: why do these need to be run on all backends?

Comment (sravfeyn) [reply testapps/test_pillowtop/tests/test_xform_pillow.py]: Perhaps not necessary for testing just heartbeat. Removing

File: corehq/apps/users/models.py (hunk 2 above)
Comment (sravfeyn) [reply corehq/apps/users/models.py]: Ah, right, thanks.

File: corehq/apps/users/models.py [hunk 4]
```diff
@@ -2856,5 +2864,51 @@ def add_sync(cls, domain, user_id, app_id, build_id, sync_date, device_id):
...
+                    build_id = EXCLUDED.build_id,
+                    sync_date = EXCLUDED.sync_date,
+                    device_id = EXCLUDED.device_id,
+                    app_version = EXCLUDED.app_version,
+                    num_unsent_forms = EXCLUDED.num_unsent_forms,
+                    num_quarantined_forms = EXCLUDED.num_quarantined_forms,
+                    commcare_version = EXCLUDED.commcare_version,
+                    build_profile_id = EXCLUDED.build_profile_id,
+                    last_heartbeat = CLOCK_TIMESTAMP()
+                WHERE staging.last_heartbeat is NULL or EXCLUDED.last_heartbeat > staging.last_heartbea
```
Comment (snopoke) [inline corehq/apps/users/models.py]:
```suggestion
                WHERE staging.last_heartbeat is NULL or EXCLUDED.last_heartbeat > staging.last_heartbeat
```

**Review by snopoke — APPROVED** [review]

**Comment by mkangia** [comment]
A happy couch! 🛋

**Comment by sravfeyn** [comment]
Yeah, much less writes on couch since its deployed on the 13th.
Screenshot 2020-01-17 at 2 11 39 PM

**Comment by mkangia** [comment]
thats great @sravfeyn 

and @emord

**Comment by dannyroberts** [comment]
@sravfeyn from the timing and context, I think this may be causing this error: https://sentry.io/organizations/dimagi/issues/1429050750/?project=136860&query=is%3Aunresolved. If you don't have time to respond on Monday, I can look into reverting to give you time to fix

**Comment by sravfeyn** [comment]
Thanks Danny.

I haven't been able to figure out the root cause. I reviewed the code again a bit it to see if the code introduces NULL values in `sync_date`, but I haven't found any code path here that would do that. Obvious fix would be to update `_last_sync_needs_update` to return `False` if sync_date is empty. But I am not 100% sure if we are not missing some other bug that's making up the sync_date to be None.

**Comment by snopoke** [comment]
@sravfeyn surely this is just from records that never had a sync date? The condition to run this code is `record.device_id or record.sync_date or record.last_heartbeat` so it's quite possible for `record.sync_date` to be `None`.

**Comment by sravfeyn** [comment]
Right, what I haven't been able to identify is whether it's coming from an actual sync or heartbeat (a submission staging update wouldn't go through this code). I haven't been able to find this particular TypeError  prior to this code was deployed for heartbeat or sync view (I searched sentry for `TypeError transaction:/a/{domain}/phone/restore/` and `typeerror transaction:/a/{domain}/phone/heartbeat/{app_build_id}/`). So, I was trying to find if I introduced any bug in this code where I am overriding actual sync-date to be `None`, but I haven't found any so far.

**Comment by sravfeyn** [comment]
This should address the issue https://github.com/dimagi/commcare-hq/pull/26445

**Pull request merged**